File: ruby/019.rb
# 019
# Remove Nth Node From End of List
# 2015-04-28
#####################################################
# Definition for singly-linked list.
class ListNode
  attr_accessor :val, :next
    def initialize(val)
      @val = val
      @next = nil
  end
end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
  q = head
  for i in 0...n
    q = q.next
  end
  if not q
    return head.next
  end
  p = head
  while q.next
    p = p.next
    q = q.next
  end
  p.next = p.next.next
  head
end

def printList(head)
  while head.next do
    print head.val, ' -> '
    head = head.next
  end
  puts head.val
end

#####################################################

a1 = ListNode.new(1)
a2 = ListNode.new(2)
a3 = ListNode.new(3)
a4 = ListNode.new(4)
a5 = ListNode.new(5)
a1.next = a2
a2.next = a3
a3.next = a4
a4.next = a5

n = 5

printList(a1)
printList(remove_nth_from_end(a1, n))

File: c/019.c
//
//  019.c
//  leetcode
//
//  Created by Hang Zhang on 4/28/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <stdio.h>
#include <stdio.h>
#include <stdlib.h>
/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode *q = head;
    for (int i = 0; i < n; i++)
        q = q->next;
    if (!q)
        return head->next;
    struct ListNode *p = head;
    while(q->next) {
        p = p->next;
        q = q->next;
    }
    p->next = p->next->next;
    return head;
}

void printList(struct ListNode* head) {
    while(head->next) {
        printf("%d -> ", head->val);
        head = head->next;
    }
    printf("%d\n", head->val);
}

int main(int arg, char *argv[]) {
    // insert code here...
    printf("LeetCode 019 Remove Nth Node From End of List, C ... ...\n");
    
    struct ListNode *a1 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *a2 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *a3 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *a4 = (struct ListNode*)malloc(sizeof(struct ListNode));
    struct ListNode *a5 = (struct ListNode*)malloc(sizeof(struct ListNode));
    a1->val = 1;
    a2->val = 2;
    a3->val = 3;
    a4->val = 4;
    a5->val = 5;
    a1->next = a2;
    a2->next = a3;
    a3->next = a4;
    a4->next = a5;
    
    int n = 5;
    
    printList(a1);
    printList(removeNthFromEnd(a1, n));
    
    return 0;
}